TDDWithURMAndKotlin/part2_increment.md

```diff
@@ -0,0 +1,157 @@
+@ValueSource(ints = [11, 15, 110, 200, Int.MAX_VALUE])
+fun `increment function adds 1 to value of position`(position: Int) {
+    zero(registry, position)
+    increment(registry, position)
+
+    assertEquals(1, registry.getValueAtPosition(position))
+}
+```
+
+To fix the breaking test we need to modify our function:
+
+```kotlin
+fun increment(registry: Registry, position: Int) {
+    require(position > 0) { "Position must be positive number" }
+
+    val currentValue = registry.getValueAtPosition(position)
+    checkNotNull(currentValue) { "Register must be initialized" }
+
+    registry.setValueAtPosition(position, currentValue + 1)
+}
+```
+
+We store the value of `registry.getValueAtPosition(position)` and then call the contract `checkNotNull` (I'll write about contracts soon, I promise!)
+
+But testing with a single `1` is boring… Let's write something that operates over different values. To test dinamycally we will need to check a parameter, then call `increment` and get an expected value. We can simply call `assertEquals(parameter + 1, registry.getValueAtPosition(parameter))` but somehow it feels like cheating…
+
+Using _parameterized tests_ from JUnit 5 we can create a single test that will cover different cases in a single method, and we will use `@CsvSource` to pass different parameters. `@CsvSource` takes a group of comma-separated-strings for every set of parameters and our method will ask for this parameters in the appropiate types:
+
+```kotlin
+@ParameterizedTest
+@CsvSource("50,1,2", "50,5,6", "50,100,101")
+fun `increment function adds 1 to any integer value`(
+    position: Int,
+    original: Int,
+    expected: Int
+){
+    registry.setValueAtPosition(position, original)
+
+    increment(registry, position)
+
+    assertEquals(expected, registry.getValueAtPosition(position))
+}
+```
+
+This allow the test to be more dynamic and we can add more cases having more flexibility!
+
+Let's write one test mixing what we learned:
+
+```kotlin
+@ParameterizedTest
+@CsvSource("10,23,24", "15,99,100", "1000,1000,1001")
+fun `check zero and increment can work together`(
+    position: Int,
+    original: Int,
+    expected: Int
+) {
+    registry.setValueAtPosition(position, original)
+    increment(registry, position)
+    assertEquals(expected, registry.getValueAtPosition(position))
+
+    zero(registry, position)
+    assertEquals(0, registry.getValueAtPosition(position))
+
+    increment(registry, position)
+    assertEquals(1, registry.getValueAtPosition(position))
+}
+```
+
+Wow! That's a test. We start by setting manually a value on a position. Then we increment it to assert the increment works as expected (we know it does because we have a test to validate it!) and then we do something "new": we reset to 0 a register. We are validating something we didn't check before, that the `zero` function can reset any registry regardless of the value. Finally we `increment` and validate that `increment` works regardless of the value.
```

Comment:
He is just describing the current test, so I think _start_ makes sense in this particular sentence.

```suggestion
Wow! That's a test. We start by manually setting a value on a position. Then we increment it to assert that the increment works as expected -we know it does because we have a test to validate it!- and then we do something "new": we reset a register to 0. Here we are validating something we didn't check before: that the `zero` function can reset any registry regardless of the value. Finally, we `increment` and validate that `increment` works regardless of the value.
```